File: 2025-2026 ICPC Bolivia Pre-National Contest/e.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int N, T;
    cin >> N >> T;

    vector<int> tiempos(N);
    for (int i = 0; i < N; ++i) {
        cin >> tiempos[i];
    }

    int mitad = N / 2;
    vector<int> grupo1(tiempos.begin(), tiempos.begin() + mitad);
    vector<int> grupo2(tiempos.begin() + mitad, tiempos.end());

    vector<int> suma_grupo1, suma_grupo2;

    int n1 = grupo1.size();
    for (int mask = 0; mask < (1 << n1); ++mask) {
        int suma = 0;
        for (int i = 0; i < n1; ++i) {
            if (mask & (1 << i)) {
                suma += grupo1[i];
            }
        }
        if (suma <= T)
            suma_grupo1.push_back(suma);
    }


    int n2 = grupo2.size();
    for (int mask = 0; mask < (1 << n2); ++mask) {
        int suma = 0;
        for (int i = 0; i < n2; ++i) {
            if (mask & (1 << i)) {
                suma += grupo2[i];
            }
        }
        if (suma <= T)
            suma_grupo2.push_back(suma);
    }

    sort(suma_grupo2.begin(), suma_grupo2.end());

    int mejor = 0;
    for (int s1 : suma_grupo1) {
        int restante = T - s1;
        auto it = upper_bound(suma_grupo2.begin(), suma_grupo2.end(), restante);
        if (it != suma_grupo2.begin()) {
            --it;
            int candidato = s1 + *it;
            if (candidato > mejor)
                mejor = candidato;
        }
    }

    cout << mejor << endl;
    return 0;
}

File: 2025-2026 ICPC Bolivia Pre-National Contest/e.py
import sys
import bisect

data = sys.stdin.read().split()
N = int(data[0])
T = int(data[1])
tiempos = list(map(int, data[2:]))

mitad = N // 2
grupo1 = tiempos[:mitad]
grupo2 = tiempos[mitad:]

suma_grupo1 = []
suma_grupo2 = []

n1 = len(grupo1)
for mask in range(1 << n1):
    s = 0
    for i in range(n1):
        if mask & (1 << i):
            s += grupo1[i]
    if s <= T:
        suma_grupo1.append(s)

n2 = len(grupo2)
for mask in range(1 << n2):
    s = 0
    for i in range(n2):
        if mask & (1 << i):
            s += grupo2[i]
    if s <= T:
        suma_grupo2.append(s)

suma_grupo2.sort()
mejor = 0
for s1 in suma_grupo1:
    restante = T - s1
    idx = bisect.bisect_right(suma_grupo2, restante)
    if idx:
        candidato = s1 + suma_grupo2[idx - 1]
        if candidato > mejor:
            mejor = candidato
print(mejor)
